### ISSUE TITLE
About landmark loss

### ISSUE THREAD

**Issue by NNNNAI** [zyainfal/One-Shot-Face-Swapping-on-Megapixels#8]
Thanks for sharing your work!
Can you tell me in detail how you did this landmark loss, which is to present the langmark in the form of a heatmap of [h,w,1], and then calculate the l2loss for the input and the generated image?
Many thanks！

**Comment by zyainfal** [comment]
You are almost right, except we use the heatmap of [h,w,c], where `c` represents the number of eye&mouth points. Then, `l2` loss is calculated for the input and the generated image.

**Comment by NNNNAI** [comment]
Thanks for your quick reply, could you tell me what the specific number of c is here, because I saw that there are several detection modes for the number of landmarks in[ High-resolution networks (HRNets)](https://github.com/HRNet/HRNet-Facial-Landmark-Detection) for facial landmark detection.

**Comment by zyainfal** [comment]
We used 300W version, which has 98 landmarks. 
And we only use later 38(c) landmarks.

**Comment by zyainfal** [comment]
We used COFW version, which has 98 landmarks.（Because I modify their source code, I'm not very sure which version I used, but I can tell that the number of landmarks is 98.）
And we only use later 38(c) landmarks.

**Comment by NNNNAI** [comment]
I see~. BTW, Have you used all 98 points for landmark loss, will the effect be better or worse? Thanks again~.

**Comment by zyainfal** [comment]
We have. 
Supervision on all 98 landmarks will change the facial shape, such as eyebrow shape and cheek shape. And this makes the swapped face look-alike the target face, so we only use the eye/mouth landmarks for training.

**Comment by NNNNAI** [comment]
Thanks for your help~. And I noticed that the HieRFE encoder get image with 256*256 as input, is that means the input image with 1024*1024 need to be reiszed as 256*256 during training?

**Comment by zyainfal** [comment]
Yes, and this requires much less GPU memory, which means a larger batchsize.

**Comment by NNNNAI** [comment]
Sorry to disturb you again, I am try to train the first step of megafaceswap(Reconstruction), I follow all the loss list in the paper except the landmark loss,cause I find the Hrnet ldm detector can only detect landmark of picture with bbox labels while the reconstruction output did not got this. I have been train 28 epoch on 4 2080ti GPU with total batch size 8. The results is shown below. It looks terrible Hhhhh.Does the lack of ldm loss casuse such a output? Would you be kind to tell me how do you get the bbox label of recontrustion output for getting the ldm loss. Many thanks~,have a nice day.
![epoch_28_step_8001_fake](https://user-images.githubusercontent.com/30747212/125148511-59d4d900-e165-11eb-9690-2dd65a414abb.jpg)

**Comment by zyainfal** [comment]
The lack of ldm loss won't cause terrible results, you may need more training epochs or more training data (e.g. images generated by stylegan version 1). 
If you still have troubles for training HieRFE, the repo of [psp encoder](https://github.com/eladrich/pixel2style2pixel) is a good baseline for you. Please note that we use the constant input `c1` of stylegan2 while they didn't. 

For the usage of HRNet, you can modify their code to drop face crop steps so that you don't need the bbox labels, as the faces in FFHQ and CelebA-HQ have been aligned already.

**Comment by NNNNAI** [comment]
Thanks for your help!! One more question about the training process .What you are saying in the paper is: Hierfe and FTM are sequentially trained seventeen epochs in total on FFHQ and the auxiliary data. Is this correct?  When training FTM, the Hierfe part is fixed, and only FTM participates in the training. Thanks for your help, have a nice day~.

**Comment by zyainfal** [comment]
Yes, you are correct. 
And you can also train them together with a learning rate of HieRFE 10x less. 
Depends on your GPU memory, both training processes should work well.

**Comment by NNNNAI** [comment]
Ok, I see. Let me try it. Many thanks~.

**Comment by NNNNAI** [comment]
I have been try to train the FTM. I used mse to calculate the value of Lnorm loss, and now the value of Lnorm itself is about 50. I set the norm weight to 100000 according to your paper, and then the value of norm_loss*norm_weight is 500W. Is this normal? . The values of other loss such as idloss and rec_loss are too small in comparison, and seem to be ineffective. When I calculate the norm value, do I need to make a l2norm for both Lshigh and ls2t and then calculate the norm loss? Many thanks~.

**Comment by zyainfal** [comment]
OMG, my apology. 
The loss weights should be 1 for reconstruction, 32 for L_norm, 32 for LPIPS, 24 for identity, and 100,000 for landmarks.

**Comment by zyainfal** [comment]
and no l2norm needed

**Comment by HardboiledHu** [comment]
> We used COFW version, which has 98 landmarks.（Because I modify their source code, I'm not very sure which version I used, but I can tell that the number of landmarks is 98.）
> And we only use later 38(c) landmarks.

I checked the code of HRNet for landmark-Detection and found that maybe the version of HRNet is WFLW, because the landmark of COFW is 29 and the number of WFLW is 98. So I want to confirm again which version it is？

**Comment by zyainfal** [comment]
Well, then use the WFLW version as long as it includes eye & mouth landmark predictions.

**Comment by seokg** [comment]
Hi, I was also wondering about the implementation details on the landmark prediction loss.

When predicting the landmark coordinate of the given input (using `decode_preds` function in [HRNet repo](https://github.com/HRNet/HRNet-Facial-Landmark-Detection/blob/master/lib/core/evaluation.py) ), the gradient is not able to pass down to the model. Is there any other way to pass the gradient to the network?

**Comment by zyainfal** [comment]
You don't need `decode_preds` function.
The landmark loss is calculated on feature maps, not the predicted coordinates.

**Comment by seokg** [comment]
Thanks for the clarification! @zyainfal

**Comment by zyainfal** [comment]
Oh wait, I just noticed that in the second stage, HieRFE trained by both swapping losses and reconstruction loss, where swapping losses make it predict better latent codes for face swapping and reconstruction loss makes it hold its fundamental function. The multi-target training gives HieRFE better generalization.
